Make median_absolute_deviation scale argument aligned w/iqr

rlucas7 opened the pull request:
#### Reference issue
Closes gh-11090 -> #11090 

#### What does this implement/fix?
In gh-11090 there was some confusion around the default value of scale in `median_absolute_deviation` function and there is also an inconsistency between how 
`median_absolute_deviation` and `iqr` accept `scale=` argument. 

In `iqr` the scale can be either a string or a float. If string it checks if lower cased is 'raw' or 'normal' with 'raw' as default which corresponds to 1.0 float value. 

This PR adds similar functionality to `median_absolute_deviation` and makes the bevavior aligned with `iqr`.

#### Additional information
Note the scaling used for `iqr` when `scale='normal'` is used is ***different*** than what is needed in `median_absolute_deviation` is a normal random variable is assumed as input (the reason for the choice of keyword) and a normalization is required to provide a consistent estimator of the standard deviation. 

Rather than duplicate the scale dictionary for `median_absolute_deviation` from `iqr`:
https://github.com/scipy/scipy/blob/a903652679ecbe787449fb99b8f3a6e105ad39b0/scipy/stats/stats.py#L2602
I chose to reuse the existing one created for `iqr` and I did a rescale for `median_absolute_deviation` if `scale='normal'` to match the loss function/use case.

rlucas7 commented:
Also, should've mentioned I'm not sure if we want to call this a signature defect and move to the default behavior that others prefer (see the linked issue) or if we want to do a full deprecation cycle. IIRC the function was released in 1.4.0 so very new.

rlucas7 commented:
Ping @tcjansen @u55 @josef-pkt as they were commenting on the related issue discussion.

u55 commented on scipy/stats/stats.py:
```diff
@@ -2789,10 +2789,15 @@ def median_absolute_deviation(x, axis=0, center=np.median, scale=1.4826,
-        ensures consistency with the standard deviation for normally distributed
-        data.
+    scale : scalar or str, optional
+        The numerical value of scale will be divided out of the final
```
This docstring does not match the code. You have to decide if you are going to keep the old behavior and multiply the result by `scale` or adopt a new behavior and divide by `scale` to match `iqr`. Right now the docstring says divide but the code does a multiplication. Similarly, the docstring says that `'normal'` equals 1.4826, but in fact the code produces scale=1/1.4826=0.6745.

rlucas7 replied:
@u55 yes, one was multiplying the other dividing. I'm working on fixing the tests now (the doctests will need to change as well), thanks for catching that.

rlucas7 replied:
Looked at this closer over the last couple days. So the point here with the scaling factor is to produce an asymptotically unbiased estimator of the standard deviation, assuming all observations are draws from a normal distribution with mean 0. To ensure that that is the case the MAD needs to either get multiplied by 1.4826 (reciprocal of the inverse cdf of std normal at 0.75) or what is equivalent, dividing by inverse cdf of std normal at 0.75. 
As we want to have alignment/consistency with `iqr` I've adjusted here to match `iqr`

rlucas7 commented on scipy/stats/stats.py:
```diff
@@ -2880,6 +2893,10 @@ def median_absolute_deviation(x, axis=0, center=np.median, scale=1.4826,
         mad = np.median(np.abs(arr - med), axis=axis)
 
+    # NOTE: _scale_conversions was for IQR so for MAD we need to divide by 2
+    if scale_key == 'normal':
```
Also I'm adding some tests and need to rethink this part, the `scale_key` isn't assigned/instantiated yet here if `scale=1.0`

rlucas7 replied:
I switched this around a bit, check only once is scale is string and handle appropriately.

rlucas7 commented on scipy/stats/stats.py:
```diff
@@ -2789,10 +2789,17 @@ def median_absolute_deviation(x, axis=0, center=np.median, scale=1.4826,
+        result. The following string values are recognized (default is 'raw'):
+
+          * 'raw' : No scaling, just return the raw sum of absolute deviations.
+          * 'normal' : Scale by :math:`\frac{1}{\sqrt{2} erf^{-1}(\frac{1}{2})}\approx 0.67449`.
```
These didn't have `*` and weren't indented correctly and we causing doctests to fail.

rlucas7 commented on scipy/stats/stats.py:
```diff
@@ -2837,23 +2844,31 @@ def median_absolute_deviation(x, axis=0, center=np.median, scale=1.4826,
     0.9973906394005013
     >>> stats.median_absolute_deviation(x)
-    1.2280762773108278
+    0.82832610097857
```
these values needed to change b/c of the change in default `scale` arg, similarly in unit tests.

rlucas7 commented on scipy/stats/tests/test_stats.py:
```diff
@@ -1890,28 +1897,28 @@ def test_mad_empty(self):
+
+    def test_mad_scale_default(self):
+        mad = stats.median_absolute_deviation(self.dat, scale="normal")
+        mad_float = stats.median_absolute_deviation(self.dat, scale=1/1.4826022185056018)
```
Here `0.67448975= 1/1.4826022185056018`, I've written this way (as 1/1.4826) to provide a link between what is commonly written (e.g. on wikipedia) and how the code does the calculation/division.

rlucas7 commented:
@u55 if you've got time and could give this another pass that would be helpful. I've tried to clarify things in docstring to reflect how the code works and align with `iqr`.

rlucas7 commented:
The only test failure (beside ppc that fails for everything) is something unrelated, 
```
/home/travis/virtualenv/python3.6.7/lib/python3.6/site-packages/scipy/sparse/linalg/tests/test_pydata_sparse.py:65: in test_isolve_gmres

    x, info = splin.gmres(A_sparse, b, atol=1e-15)

<decorator-gen-9>:2: in gmres
```

u55 commented:
I don't see any obvious errors now. Does the docstring need a warning that the default value of `scale` has changed? What about adding an entry to the changelog?

rlucas7 commented:
>> On Feb 13, 2020, at 3:02 AM, u55 <notifications@github.com> wrote:
> ﻿
> I don't see any obvious errors now. Does the docstring need a warning that the default value of scale has changed? What about adding an entry to the changelog?
> 
The docstring says that the default is raw, division by 1. 

We usually add the notes entry to the wili change log when the PR is merged so I’ll hold off for now. 

The warning that the docstring has changed. Usually that sort of changes goes through a deprecation cycle which consists of 2 or more releases maintaining old behavior. Given that the function was just added and there seems to be agreement that the previous default is non-obvious, I’m considering that a defect and this fixes the defect. Defects don’t go through deprecation cycle. 

However, if you or others on the scipy-dev team prefer We can go through a deprecation cycle that persists existing behavior for now and changes in 1.7 release in another year.

> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.

rlucas7 commented:
just updated my upstream against master and rebased, hoping to get the ppc fix that Tyler put in earlier and hopefully get this PR all green.

rgommers commented:
@rlucas7 if we go with `median_abs_deviation(..., scale='raw')` as I just proposed in gh-11090, I think you can keep this PR open, rename the current function, and re-add `median_absolute_deviation = median_absolute_deviation(scale='normal')` and decorate it with `np.deprecate`.

rlucas7 commented:
> @rlucas7 if we go with `median_abs_deviation(..., scale='raw')` as I just proposed in [gh-11090](https://github.com/scipy/scipy/issues/11090), I think you can keep this PR open, rename the current function,

ok. This sounds like the best course of action. 

> and re-add `median_absolute_deviation = median_absolute_deviation(scale='normal')` and decorate it with `np.deprecate`.

I'm not sure I follow this statement, do you mean to create a partially applied function using the new (from your proposal above) function? IIUC that would be something like: 

```
from functools import partial
median_absolute_deviation = partial(median_abs_deviation, scale='normal')
```

or do you want to also keep the previously written function, `median_absolute_deviation`?

rlucas7 commented:
> > and re-add `median_absolute_deviation = median_absolute_deviation(scale='normal')` and decorate it with `np.deprecate`.
> 
> I'm not sure I follow this statement, do you mean to create a partially applied function using the new (from your proposal above) function? IIUC that would be something like:
> 
> ```
> from functools import partial
> median_absolute_deviation = partial(median_abs_deviation, scale='normal')
> ```
> 
> or do you want to also keep the previously written function, `median_absolute_deviation`?

I tried the approach I'm suggesting here, it works locally, not sure how the `partial` will work with 
the doctests and other stuff in CI so pushing to see if it breaks. 

@rgommers let me know if you had something else in mind. This change will still need some cleanup, I'll go over it again next weekend if you agree on the approach espoused in latest changeset (and nothing breaks in CI).

rgommers commented:
I didn't mean anything that fancy, simple:
```
@np.deprecate(...)   # generates the right deprecation warning
def median_absolute_deviation(...):
    return median_absolute_deviation(..., scale='normal')
```

rlucas7 commented:
All the travis builds pass except one that timed out on setup:
https://travis-ci.org/github/scipy/scipy/jobs/664981743#L298
not clear to me what the error is, it appears to be something in the sphinx build, when `make` is called in a subprocess call:
https://circleci.com/gh/scipy/scipy/17879?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link
(scroll to bottom)

rlucas7 commented:
> I didn't mean anything that fancy, simple:
> 
> ```
> @np.deprecate(...)   # generates the right deprecation warning
> def median_absolute_deviation(...):
>     return median_abs_deviation(..., scale='normal')
> ```

Just checking one thing, I think the arguments should be passed in to the `median_abs_deviation`
should be the defaults that are used int the `median_absolute_deviation`, right? 

If I understood correctly, then I think the default should be not:

>     return median_abs_deviation(..., scale='normal')

rather: 

>     return median_abs_deviation(..., scale=1.4826)

right? 

I'm assuming we ***don't*** want to change the default behavior for the `median_absolute_deviation` function `scale` argument but ***do*** want to change to `'raw'` as default scale in `median_abs_deviation`. 

@rgommers do I understand correctly?

rlucas7 commented:
The only fail in the travis builds is the ppc fail on a timeout building the env. 

ref: https://travis-ci.org/github/scipy/scipy/jobs/670797621#L552,L558

everything else green.

rgommers commented:
> I'm assuming we _**don't**_ want to change the default behavior for the `median_absolute_deviation` function `scale` argument but _**do**_ want to change to `'raw'` as default scale in `median_abs_deviation`.
> 
> @rgommers do I understand correctly?

Yes that's correct.

rgommers commented on scipy/stats/stats.py:
```diff
@@ -2735,9 +2737,9 @@ def iqr(x, axis=None, rng=(25, 75), scale='raw', nan_policy='propagate',
 
-def median_absolute_deviation(x, axis=0, center=np.median, scale=1.4826,
+def median_abs_deviation(x, axis=0, center=np.median, scale='raw',
                               nan_policy='propagate'):
```
minor: indentation is off here

rlucas7 replied:
fixed.

rlucas7 replied:
there were a few other spots with alignment errors (`E128`s) and `E501`s and which I noticded when I ran pycodestyle, so I fixed those too.

rgommers commented on scipy/stats/__init__.py:
```diff
@@ -201,6 +201,7 @@
    mvsdist
    entropy
    median_absolute_deviation
```
you can remove this, best to only list the non-deprecated function

rgommers replied:
you may then need to add it to the skiplist in `refguide_check.py` (not sure, could be it skips deprecated functions already)

rlucas7 replied:
Removed. 

> you may then need to add it to the skiplist in `refguide_check.py` (not sure, could be it skips deprecated functions already)

It looks like `refguide_check.py` skips them already: 
this function splits them out (deprecated are determined by executing with empty dict and checking if a deprecated warning is thrown), 
https://github.com/scipy/scipy/blob/c1372d8aa90a73d8a52f135529293ff4edb98fc8/tools/refguide_check.py#L877 
then two lines here:
https://github.com/scipy/scipy/blob/c1372d8aa90a73d8a52f135529293ff4edb98fc8/tools/refguide_check.py#L881-L882
the first does a count of deprecated and not and the second does a set difference, removing the deprecated, so we should be good here, unless I missed something :).

rlucas7 commented:
The only test failure is the ppc build in travis which failed on a time limit exceeded:
https://travis-ci.org/github/scipy/scipy/jobs/671900747#L688

@rgommers would you have a minute to take another look? 

I addressed your commented and fixed a couple additional pep8 issues by `mv`-ing out the `tox.ini` file and then moving back so that `pycodestyle` would give me the `E501` and `E128` errors.

WarrenWeckesser commented:
@rlucas7, thanks for working on this.

The advantage of creating a new function, `median_abs_deviation`, is that we start with a clean slate, API-wise.

Initially I was going to suggest removing the `'raw'` option, and just make the default `scale=1`.  (I think it was a mistake to include the `'raw'` option in `iqr`, and I don't think we should repeat that mistake here.)  But instead, I have  a more radical proposal: let's just remove the `scale` argument from `median_abs_deviation`.  Think about it:
* the `scale` argument is a trivial capability;
* the only two use-cases that have been discussed are `scale=1` and `scale=1.4826...`.  The latter can be addressed in the docstring, e.g. add something like this to the beginning or in the "Notes":

      To compute a consistent estimator of the standard deviation of a sample
      from the normal distribution, divide the MAD by the 0.75-quantile of the
      normal distribution.

  and in the "Examples", we add something like this at the end:

      Given that `x` is a sample from the normal distribution, we can use the MAD
      as a consistent estimator of the standard deviation:

      >>> from scipy.special import ndtri
      >>> stats.median_abs_deviation(x) / ndtri(0.75)

If it turns out that we get a lot of demand for a `scale` argument, we can easily add it later.  For now, I think it is better to keep the API of the new function as simple as possible.

We would still deprecate `median_absolute_deviation`.  But the deprecation shouldn't remove  the docstring of `median_absolute_deviation` like it does now, e.g. entering `median_absolute_deviation?` in an ipython shell should still show the docstring of the deprecated function.

rlucas7 commented:
@WarrenWeckesser You proposal has some merits. I agree that the scale is trivial. I'm reluctant to change the signature though. The argument for the `scale` default being incorrect and not 'raw' is in reference to the `iqr` function and being internally consistent within the `stats ` sub-package. So making the change you are proposing would really be going against the motivation for the deprecation, unless the `iqr` was also deprecated which wasn't really in the scope of this PR. 

My (current) opinion would be to support normal distribution scaling as that is a common distribution and many convergence modes lead to normal (e.g. central limit theorems). Also, ***not*** support scaling by any other distribution constants, those can be handled by user code as you mention they're usually minor changes to implement provided user has code to calculate the distributional constant.

If other's agreed to deprecate the `scale` functionality in `iqr` in coordination with this deprecation here I wouldn't be opposed to the proposal but for now I'm a no vote on the change but not firmly no.

rlucas7 commented:
>We would still deprecate median_absolute_deviation. But the deprecation shouldn't remove the docstring of median_absolute_deviation like it does now, e.g. entering median_absolute_deviation? in an ipython shell should still show the docstring of the deprecated function.

I think the overwrite of that docstring is the behavior of the `@np.deprecate` annotation which we seem to use regularly for deprecations. I'd think that putting the deprecation notice in the doc is what we want to do, notify people of future potentially breaking changes. 

Maybe I am misunderstanding this docstring comment?

WarrenWeckesser commented:
@rlucas7 wrote

> My (current) opinion would be to support normal distribution scaling as that is a common distribution and many convergence modes lead to normal (e.g. central limit theorems).

OK, but let's get rid of the `'raw'` option and make the default `scale=1`.  The additional string `'raw'` is unnecessary.   It is just one more thing that we have to explain in the docstring, and it means exactly the same thing as scale=1.


> I think the overwrite of that docstring is the behavior of the @np.deprecate annotation which we seem to use regularly for deprecations.

We definitely want it known that the function is deprecated, but we don't want to completely remove the old docstring.  The `@np.deprecate` decorator inserts the text about deprecation into the existing docstring.  Take a look at, for example, `itemfreq`:

https://github.com/scipy/scipy/blob/8a6ac2cfb353dcd5be3edca9645ecf4bd32a9c23/scipy/stats/stats.py#L1668

and see how it looks on the web: https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.itemfreq.html

rlucas7 commented:
> @rlucas7 wrote
> 
> > My (current) opinion would be to support normal distribution scaling as that is a common distribution and many convergence modes lead to normal (e.g. central limit theorems).
> 
> OK, but let's get rid of the `'raw'` option and make the default `scale=1`. The additional string `'raw'` is unnecessary. It is just one more thing that we have to explain in the docstring, and it means exactly the same thing as scale=1.

If I understand what you're suggesting correctly that means we'd removed the string argument functionality entirely and that wouldn't be consistent with the api consistency between this method and `iqr`. @WarrenWeckesser Could you please clarify what you mean here?

> > I think the overwrite of that docstring is the behavior of the @np.deprecate annotation which we seem to use regularly for deprecations.
> 
> We definitely want it known that the function is deprecated, but we don't want to completely remove the old docstring. The `@np.deprecate` decorator inserts the text about deprecation into the existing docstring. Take a look at, for example, `itemfreq`:
> 
> https://github.com/scipy/scipy/blob/8a6ac2cfb353dcd5be3edca9645ecf4bd32a9c23/scipy/stats/stats.py#L1668
> 
> and see how it looks on the web: https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.itemfreq.html

I'm guessing this is because of the partial function approach? Not clear if the `functools.partial` had similar behavior or if that one also dropped the docstring. 

I could append a custom docstring to the existing one to get what we need without writing
the entire function, e.g. something like 

```
old_doc = " the old docstring that we wanted to keep but is getting dropped by the partial function approach"
median_absolute_deviation.__doc__ + old_doc
```
which AFAICS should work but not sure if that would break anything in sphinx. 
Unless I hear otherwise I'll give that a go this weekend.

WarrenWeckesser commented:
> If I understand what you're suggesting correctly that means we'd removed the string argument functionality entirely and that wouldn't be consistent with the api consistency between this method and iqr.

My *preference* would be to eliminate the string handling entirely, but I'm OK with accepting the string `'normal'` based on your earlier comments.  What I am proposing is that the string `'raw'` not be accepted.  Instead, make the default scale 1.

As far as consistency with `iqr` goes, I'd be happy to change the default value of its `scale` argument from `'raw'` to 1, and deprecate the use of `'raw'`.   Because `scale='raw'` means the same thing as `scale=1`, we could make that change without breaking backwards compatibility. The `'raw'` option is not any more convenient than 1, it doesn't convey any more information than the numerical value--in fact, it forces the user to double-check the meaning of `'raw'`.  One could say it makes the API "bigger"--more stuff to support and explain--with no clear benefit.

WarrenWeckesser commented:
> I could append a custom docstring to the existing one to get what we need without writing
the entire function, e.g. something like

    old_doc = " the old docstring that we wanted to keep but is getting dropped by the partial function approach"
    median_absolute_deviation.__doc__ + old_doc

That shouldn't be necessary.  Have you tried just decorating the function with `@deprecate` and not removing the old docstring from the function definition?   The decorator should take care of adding text to the beginning of the docstring about it being deprecated.

rlucas7 commented:
> > I could append a custom docstring to the existing one to get what we need without writing
> > the entire function, e.g. something like
> 
> ```
> old_doc = " the old docstring that we wanted to keep but is getting dropped by the partial function approach"
> median_absolute_deviation.__doc__ + old_doc
> ```
> 
> That shouldn't be necessary. Have you tried just decorating the function with `@deprecate` and not removing the old docstring from the function definition? The decorator should take care of adding text to the beginning of the docstring about it being deprecated.

Hmm, maybe I'm not communicating the issue clearly.  The old function was: `median_absolute_deviation` the new one is `median_abs_deviation`. We've made the new one by renaming and changing the default value for scale. The old function `median_absolute_deviation` is just a pass through return of the new function with the given old default argument. 

I initially had it as a `functools.partial`, see e.g. 
https://docs.python.org/3.6/library/functools.html
Suggestion was to do the partial without the functools, so I changed. 

However, I looked at it some more and it looks like `functools.partial` would also drop the docstring in the partial function.

Hopefully that clarifies. It might work to just put the docstring into the partially applied function, ***before*** the call like so, 
```
def median_absolute_deviation(x, axis=0, center=np.median, scale=1.4826,
                              nan_policy='propagate'):
""" docstring for median_absolute_deviation here"""
    return median_abs_deviation(x, axis=axis, center=center, scale=scale,
                                nan_policy=nan_policy)
```

rlucas7 commented:
> Hopefully that clarifies. It might work to just put the docstring into the partially applied function, _**before**_ the call like so,
> 
> ```
> def median_absolute_deviation(x, axis=0, center=np.median, scale=1.4826,
>                               nan_policy='propagate'):
> """ docstring for median_absolute_deviation here"""
>     return median_abs_deviation(x, axis=axis, center=center, scale=scale,
>                                 nan_policy=nan_policy)
> ```

That worked for me locally to include the old docstring and deprecation warning.

rlucas7 commented:
> > If I understand what you're suggesting correctly that means we'd removed the string argument functionality entirely and that wouldn't be consistent with the api consistency between this method and iqr.
> 
> My _preference_ would be to eliminate the string handling entirely, but I'm OK with accepting the string `'normal'` based on your earlier comments. What I am proposing is that the string `'raw'` not be accepted. Instead, make the default scale 1.
> 
> As far as consistency with `iqr` goes, I'd be happy to change the default value of its `scale` argument from `'raw'` to 1, and deprecate the use of `'raw'`. Because `scale='raw'` means the same thing as `scale=1`, we could make that change without breaking backwards compatibility. The `'raw'` option is not any more convenient than 1, it doesn't convey any more information than the numerical value--in fact, it forces the user to double-check the meaning of `'raw'`. One could say it makes the API "bigger"--more stuff to support and explain--with no clear benefit.

I changed the scale default to `1.0`. For the 'raw' deprecation, the current implementation uses the same dictionary as `iqr` so I guess we'll need to include that in here as well even though I wanted to separate the two changesets. 
I'll keep the `'raw'` string support but issue a warning that the use of raw is deprecated. Both in `median_abs_deviation` and in `iqr`.

rlucas7 commented:
Test failures look unrelated:
scipy/linalg/tests/test_lapack.py:2189: in test_pttrf_pttrs

tylerjereddy commented:
@rlucas7 @WarrenWeckesser I see this is still active--are you likely to reach agreement/merge readiness in about a week for the branching of 1.5.x?

rlucas7 commented:
> @rlucas7 @WarrenWeckesser I see this is still active--are you likely to reach agreement/merge readiness in about a week for the branching of 1.5.x?

Thanks for the nudge @tylerjereddy it looks like a couple linter errors I need to fix then hopefully only minor changes but I'll let @WarrenWeckesser comment on readiness

rlucas7 commented:
@WarrenWeckesser just did a rebase with changes based on discussion. I switched out `iqr` to use `scale=1.0` as default, emit a `DeprecationWarning` when  `scale='raw'` is used in either `iqr` or in `median_abs_deviation` and fixed the dropped docstring issue in the (deprecated) `median_absolute_deviation`. Please take a look and let me know if further changes required. I can do them this weekend if needed if we're ok w/these changes into 1.5.0 branch.

rlucas7 commented:
Not sure why the circleci builds failed but everything else is green :)

larsoner commented:
CircleCI treats warnings as errors (except for some that we permit), so if you search the output for "warning" you should find some relevant to your changes:
```
/home/circleci/repo/venv/lib/python3.7/site-packages/numpy/lib/utils.py:docstring of scipy.stats.median_absolute_deviation:20: WARNING: Unexpected indentation.
/home/circleci/repo/venv/lib/python3.7/site-packages/numpy/lib/utils.py:docstring of scipy.stats.median_absolute_deviation:21: WARNING: Block quote ends without a blank line; unexpected unindent.
/home/circleci/repo/venv/lib/python3.7/site-packages/numpy/lib/utils.py:docstring of scipy.stats.median_absolute_deviation:35: WARNING: Unexpected indentation.
/home/circleci/repo/venv/lib/python3.7/site-packages/numpy/lib/utils.py:docstring of scipy.stats.median_absolute_deviation:38: WARNING: Definition list ends without a blank line; unexpected unindent.
/home/circleci/repo/venv/lib/python3.7/site-packages/numpy/lib/utils.py:docstring of scipy.stats.median_absolute_deviation:39: WARNING: Unexpected section title.

Returns
-------
/home/circleci/repo/venv/lib/python3.7/site-packages/numpy/lib/utils.py:docstring of scipy.stats.median_absolute_deviation:45: WARNING: Definition list ends without a blank line; unexpected unindent.
/home/circleci/repo/venv/lib/python3.7/site-packages/numpy/lib/utils.py:docstring of scipy.stats.median_absolute_deviation:46: WARNING: Unexpected section title.

See Also
--------
/home/circleci/repo/venv/lib/python3.7/site-packages/numpy/lib/utils.py:docstring of scipy.stats.median_absolute_deviation:79: WARNING: Unexpected indentation.
/home/circleci/repo/venv/lib/python3.7/site-packages/numpy/lib/utils.py:docstring of scipy.stats.median_absolute_deviation:80: WARNING: Block quote ends without a blank line; unexpected unindent.
...
/home/circleci/repo/venv/lib/python3.7/site-packages/numpy/lib/utils.py:docstring of scipy.stats.median_absolute_deviation:11: WARNING: citation not found: Rdfe996141950-1
/home/circleci/repo/venv/lib/python3.7/site-packages/numpy/lib/utils.py:docstring of scipy.stats.median_absolute_deviation:11: WARNING: citation not found: Rdfe996141950-2
```

larsoner commented on scipy/stats/stats.py:
```diff
@@ -2760,12 +2767,18 @@ def median_absolute_deviation(x, axis=0, center=np.median, scale=1.4826,
+        The numerical value of scale will be divided out of the final
+        result. The following string value is also recognized (default is 1.0):
+
+          * 'normal' : Scale by :math:`\frac{1}{\sqrt{2}erf^{-1}(1/2)}`.
```
I would unindent this by two spaces, otherwise it will appear as a bullet inside a block quote instead of just a bullet

rlucas7 replied:
done

larsoner commented on scipy/stats/stats.py:
```diff
@@ -2858,13 +2893,111 @@ def median_absolute_deviation(x, axis=0, center=np.median, scale=1.4826,
+    absolute deviations from the median. It is a measure of dispersion
+    similar to the standard deviation but more robust to outliers [2]_.
+    The MAD of an empty array is ``np.nan``.
+    .. versionadded:: 1.3.0
```
Needs nowline before and after

rlucas7 replied:
yes, done

larsoner commented on scipy/stats/stats.py:
```diff
@@ -2858,13 +2893,111 @@ def median_absolute_deviation(x, axis=0, center=np.median, scale=1.4826,
+def median_absolute_deviation(x, axis=0, center=np.median, scale=1.4826,
+                              nan_policy='propagate'):
+    """
+    Compute the median absolute deviation of the data along the given axis.
```
Needs newline after first line

rlucas7 replied:
done

larsoner commented on scipy/stats/stats.py:
```diff
@@ -2858,13 +2893,111 @@ def median_absolute_deviation(x, axis=0, center=np.median, scale=1.4826,
+        The following options are available (default is 'propagate'):
+          * 'propagate': returns nan
+          * 'raise': throws an error
+          * 'omit': performs the calculations ignoring nan values
```
Needs newline, etc -- lots of newlines missing in this docstring

rlucas7 replied:
yes, I see now, before `* 'propagate'` line

rlucas7 commented:
There are a bunch of other deprecation warning emitted from the doc builds that are unrelated. 
I'm not sure if those also cause the build to fail, they are mostly around 'frechet_{l,r}' deprecation. 

The last 2 I'm seeing are: 
```
/home/circleci/repo/venv/lib/python3.7/site-packages/numpy/lib/utils.py:docstring of scipy.stats.median_absolute_deviation:24: WARNING: Block quote ends without a blank line; unexpected unindent.
/home/circleci/repo/venv/lib/python3.7/site-packages/numpy/lib/utils.py:docstring of scipy.stats.median_absolute_deviation:65: WARNING: Error in "rubric" directive:
no content permitted.

.. rubric:: References

    .. [Rdfe996141950-1] "Median absolute deviation" https://en.wikipedia.org/wiki/Median_absolute_deviation
    .. [Rdfe996141950-2] "Robust measures of scale" https://en.wikipedia.org/wiki/Robust_measures_of_scale
```
Not sure but I think the last one can be fixed by a newline between the two refs. 
I'm stymied by the first one though, from looking at line count difference it looks like one of the lines in the `Parameters` section is off but nothing looks awry to me, @larsoner do you know what is the doc build error here?

larsoner commented on scipy/stats/stats.py:
```diff
@@ -2858,13 +2893,124 @@ def median_absolute_deviation(x, axis=0, center=np.median, scale=1.4826,
+    ----------
+    x : array_like
+        Input array or object that can be converted to an array.
+    axis : int or None, optional
```
If you look at the output, it's broken:

https://19378-1460385-gh.circle-artifacts.com/0/html-scipyorg/generated/scipy.stats.median_absolute_deviation.html#scipy.stats.median_absolute_deviation

My guess is that it's the newlines in `_median_absolute_deviation_deprec_msg`. I would try getting rid of those and doing `make html-scipyorg` in `doc/` locally and check to see if it fixes things. If it doesn't, next I would remove the `np.deprecate` altogether to rule that out.

Sometimes there are errant empty unicode spaces that can mess things up if you did a copy-paste from rendered GitHub HTML (?), but I pulled your branch locally and don't see them so I doubt that's it.

I can look tomorrow on my machine where I can actually properly build SciPy if you can't figure it out in the meantime.

larsoner replied:
`np.deprecate` is the problem:

https://github.com/numpy/numpy/pull/16349

I think we can work around this issue by unindenting the whole docstring. I'll try and push if it works.

WarrenWeckesser commented:
> @WarrenWeckesser just did a rebase with changes based on discussion. I switched out iqr to use scale=1.0 as default, emit a DeprecationWarning when scale='raw' is used...

Great, thanks.

>  ... in either iqr or in median_abs_deviation...

Let's not do that for `median_abs_deviation`.  This is a new function.  We shouldn't have any deprecated behavior built in to a new function!  I realize you are trying to share code/implementation details with `iqr`, but given that `raw` is deprecated in `iqr`, it doesn't make sense to try to push the code-sharing so far.

WarrenWeckesser commented:
The handling of `scale` can be simplified to

```
    if isinstance(scale, str):
        if scale.lower() == 'normal':
            scale = 0.6744897501960817  # special.ndtri(0.75)
        else:
            raise ValueError(f"{scale} is not a valid scale value.")
```
Also remove the code near the end of the function that is commented by `# NOTE:` that divides `scale` by 2.

WarrenWeckesser commented:
Testing this locally exposed a couple pre-existing bugs.  If you are interested, it would be nice to fix them here, but it could also be done in a follow-up.

* These lines

  ```
      if contains_nan and nan_policy == 'propagate':
          return np.nan
  ```

  do not handle the `axis` argument correctly.  E.g.

  ```
  In [48]: w
  Out[48]: 
  array([[ 1.,  2., nan,  3.,  4.],
         [ 1.,  4.,  5.,  8.,  9.]])
  
  In [49]: median_abs_deviation(w, axis=1)
  Out[49]: nan
  ```
  The return value should be `array([nan, 3.0])`.

* This use of `masked_invalid(x)`
  ```
          arr = ma.masked_invalid(x).compressed()
  ```
  is not the correct way to handle `nan_policy='omit'` because it also masks `inf`.  For example, the following two calls of `median_abs_deviation` should give the same result:

  ```
  In [59]: z = np.array([1, 3, 4, 5, 99, np.nan, np.inf])
  
  In [60]: median_abs_deviation(z, nan_policy='omit')                                                                                 
  Out[60]: 1.0
  
  In [61]: median_abs_deviation(z[~np.isnan(z)]) 
  Out[61]: 2.5
  ```
  (See https://github.com/scipy/scipy/pull/12069.)

rlucas7 commented:
> The handling of `scale` can be simplified to
> 
> ```
>     if isinstance(scale, str):
>         if scale.lower() == 'normal':
>             scale = 0.6744897501960817  # special.ndtri(0.75)
>         else:
>             raise ValueError(f"{scale} is not a valid scale value.")
> ```
> 
> Also remove the code near the end of the function that is commented by `# NOTE:` that divides `scale` by 2.



> > @WarrenWeckesser just did a rebase with changes based on discussion. I switched out iqr to use scale=1.0 as default, emit a DeprecationWarning when scale='raw' is used...
> 
> Great, thanks.
> 
> > ... in either iqr or in median_abs_deviation...
> 
> Let's not do that for `median_abs_deviation`. This is a new function. We shouldn't have any deprecated behavior built in to a new function! I realize you are trying to share code/implementation details with `iqr`, but given that `raw` is deprecated in `iqr`, it doesn't make sense to try to push the code-sharing so far.

:) yeah looks like I got a bit overzealous in my deprecations :) 

I've removed the deprecate warning emission in `median_abs_deviation`
and throw only in `median_absolute_deviation` (and `iqr`). 

> The handling of `scale` can be simplified to
> 
> ```
>     if isinstance(scale, str):
>         if scale.lower() == 'normal':
>             scale = 0.6744897501960817  # special.ndtri(0.75)
>         else:
>             raise ValueError(f"{scale} is not a valid scale value.")
> ```
> 
> Also remove the code near the end of the function that is commented by `# NOTE:` that divides `scale` by 2.

Both Done. 

>Testing this locally exposed a couple pre-existing bugs. If you are interested, it would be nice to fix them here, but it could also be done in a follow-up.

@WarrenWeckesser  I'm pushing changes without the fixes to those bugs. I'm looking at the changes required (and writing proper coverage unit tests) doesn't look like I'll have enough time over the coming weekend to do the issue justice before Tyler branches 1.5.0. If we want to remove the 1.5.0 milestone I'll keep working on it here, otherwise I can do it in a follow up PR. Let me know which you prefer. 

and thanks to @larsoner for the docstring fixes!

WarrenWeckesser commented on scipy/stats/tests/test_stats.py:
```diff
@@ -1875,65 +1875,95 @@ def test_zscore_nan_raise(self):
-        mad = stats.median_absolute_deviation(dat, axis=0)
-        mad_expected = np.asarray([0.644931, 0.7413, 0.66717, 0.59304])
+    def test_median_abs_deviation(self):
+        assert_almost_equal(stats.median_abs_deviation(self.dat, axis=None), 0.355)
```
Line length:
```suggestion
        assert_almost_equal(stats.median_abs_deviation(self.dat, axis=None),
                            0.355)
```
I realize this was already too long before being changed, but when modifying existing code, we might as well also fix PEP-8 issues.

WarrenWeckesser commented on scipy/stats/tests/test_stats.py:
```diff
@@ -1875,65 +1875,95 @@ def test_zscore_nan_raise(self):
-        mad = stats.median_absolute_deviation(dat, nan_policy='propagate')
+        with suppress_warnings() as sup:
+            sup.filter(DeprecationWarning)
+            mad = stats.median_absolute_deviation(self.dat_nan, nan_policy='propagate')
```
Line length:
```suggestion
            mad = stats.median_absolute_deviation(self.dat_nan,
                                                  nan_policy='propagate')
```

WarrenWeckesser commented on scipy/stats/tests/test_stats.py:
```diff
@@ -1875,65 +1875,95 @@ def test_zscore_nan_raise(self):
+        with suppress_warnings() as sup:
+            sup.filter(DeprecationWarning)
+            mad = stats.median_absolute_deviation(self.dat, scale="normal")
+            mad_float = stats.median_absolute_deviation(self.dat, scale=1/1.4826022185056018)
```
```suggestion
            scale = 1/1.4826022185056018
            mad_float = stats.median_absolute_deviation(self.dat, scale=scale)
```

rlucas7 commented on scipy/stats/stats.py:
```diff
@@ -2858,13 +2888,126 @@ def median_absolute_deviation(x, axis=0, center=np.median, scale=1.4826,
+        )
+        scale = 1.0
+    return median_abs_deviation(x, axis=axis, center=center, scale=scale,
+                                nan_policy=nan_policy)
```
done.

rlucas7 commented on scipy/stats/stats.py:
```diff
@@ -2858,13 +2888,126 @@ def median_absolute_deviation(x, axis=0, center=np.median, scale=1.4826,
+2.9652
+"""
+    if isinstance(scale, str) and scale.lower() == 'raw':
+        warnings.warn(
+        "use of scale='raw' is deprecated, use scale=1.0 instead",
+        DeprecationWarning
```
hmm, based on what was mentioned before I removed this functionality from `median_abs_deviation`, did that changeset not get displayed in github?

WarrenWeckesser replied:
Ah, sorry.  My comment about the change should be "Same comments as for `iqr`:"

WarrenWeckesser commented on scipy/stats/stats.py:
```diff
@@ -2711,6 +2713,11 @@ def iqr(x, axis=None, rng=(25, 75), scale='raw', nan_policy='propagate',
         if scale_key not in _scale_conversions:
             raise ValueError("{0} not a valid scale for `iqr`".format(scale))
+        if scale_key == 'raw':
+            warnings.warn(
+            "use of scale='raw' is deprecated, use scale=1.0 instead",
+            DeprecationWarning
```
This is still not fixed in ee320a82b0d; note that this is the warning in `iqr`, not `median_absolute_deviation`.

rlucas7 replied:
> This is still not fixed in [ee320a8](https://github.com/scipy/scipy/commit/ee320a82b0dfa683bb6e7b5f41393d9e8a10a23c); note that this is the warning in `iqr`, not `median_absolute_deviation`.

ah yes, I thought that was referring to `median_abs_deviation`, I've changed in `iqr` with latest changset

WarrenWeckesser commented on scipy/stats/stats.py:
```diff
@@ -2858,13 +2889,134 @@ def median_absolute_deviation(x, axis=0, center=np.median, scale=1.4826,
+# Keep the top newline so that the message does not show up on the stats page
+_median_absolute_deviation_deprec_msg = """
+To preserve the existing behavior, use
+`scipy.stats.median_abs_deviation(..., scale=1.4826)`.
```
I think this message should now show `scale=1/1.4826`.

Here's scipy 1.4.1:
```
In [77]: stats.median_absolute_deviation([1, 2, 4, 5, 9])
Out[77]: 2.9652
```
To get that same behavior with `median_abs_deviation`, I need to use `1/1.4826`:
```
In [17]: stats.median_abs_deviation([1, 2, 4, 5, 9], scale=1/1.4826)
Out[17]: 2.9652
```

WarrenWeckesser replied:
And perhaps further clarify the intent of this message by add "default" in the previous line, so it says
```
To preserve the existing default behavior, use
```

rlucas7 replied:
yes, done on both counts

WarrenWeckesser commented on scipy/stats/tests/test_stats.py:
```diff
@@ -1875,65 +1875,99 @@ def test_zscore_nan_raise(self):
 
-        mad = stats.median_absolute_deviation(dat, nan_policy='omit')
-        assert_almost_equal(mad, 0.504084)
+    def test_mad_scale_default(self):
```
Rename this test; this name is already used in line 1955.  (`flake8` caught this.)

rlucas7 replied:
renamed to `test_mad_scale_normal` and rebased the branch in latest (force) push

WarrenWeckesser replied:
+1 for rebasing and squashing the commits.  Doing that is the perfect time to also clean up the commit message.  We no longer need that itemized list of commit messages for individual commits that no longer exist.  Can you simplify the commit message so that it only explains what actually changed in the final version of the PR?

WarrenWeckesser replied:
I'll update the commit message when I merge this.  The docstrings need a couple more small tweaks, which I'm working on now.  Unfortunately, I'm still not able to build the docs locally, so I'll have to check the result generated by CircleCI.

WarrenWeckesser commented on scipy/stats/stats.py:
```diff
@@ -2686,7 +2687,8 @@ def iqr(x, axis=None, rng=(25, 75), scale='raw', nan_policy='propagate',
           'raw' : No scaling, just return the raw IQR.
           'normal' : Scale by :math:`2 \sqrt{2} erf^{-1}(\frac{1}{2}) \approx 1.349`.
```
```suggestion
          * 'raw' : No scaling, just return the raw IQR.
            **Deprecated!**  Use `scale=1` instead.
          * 'normal' : Scale by
            :math:`2 \sqrt{2} erf^{-1}(\frac{1}{2}) \approx 1.349`.
```

WarrenWeckesser commented on scipy/stats/stats.py:
```diff
@@ -2834,19 +2841,25 @@ def median_absolute_deviation(x, axis=0, center=np.median, scale=1.4826,
+        np.median. Any user defined function used will need to have the
+        function signature ``func(arr, axis)``.
+    scale : scalar or str, optional
+        The numerical value of scale will be divided out of the final
+        result. The following string value is also recognized (default is 1.0):
+
+        * 'normal' : Scale by :math:`\frac{1}{\sqrt{2}erf^{-1}(1/2)}`.
+
+        Array-like scale is also allowed, as long
+        as it broadcasts correctly to the output such that
+        ``out / scale`` is a valid operation. The output dimensions
+        depend on the input array, `x`, and the `axis` argument.
```
```suggestion
        The numerical value of scale will be divided out of the final
        result. The default is 1.0. The string "normal" is also accepted,
        and results in `scale` being the inverse of the standard normal
        quantile function at 0.75, which is approximately 0.67449.
        Array-like scale is also allowed, as long as it broadcasts correctly
        to the output such that ``out / scale`` is a valid operation. The
        output dimensions depend on the input array, `x`, and the `axis`
        argument.
```

WarrenWeckesser commented:
AppVeyor might have gotten confused by the two new commits. Closing and reopening to rerun.

WarrenWeckesser commented:
After tweaking the docstrings, the CircleCI artifacts looked good and all tests passed, so merged.

I updated the release notes with the changes made here.

Thanks @rlucas7, all.

rlucas7 commented:
Just seeing now, thanks for getting this in in time Warren.

Sincerely, 

-Lucas Roberts 

> On May 24, 2020, at 11:49 PM, Warren Weckesser <notifications@github.com> wrote:
> 
> ﻿
> After tweaking the docstrings, the CircleCI artifacts looked good and all tests passed, so merged.
> 
> I updated the release notes with the changes made here.
> 
> Thanks @rlucas7, all.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.
